library(tidyverse) #ok
library(gplots)
library(magrittr)
library(forcats)
library(vcfR)
library(reshape2)
library(FactoMineR)
library(factoextra)
library(ape)
library(RColorBrewer)
library(poolfstat)
library(stringr)
library(vroom)
library(reshape2)
library(pcadapt)
library(gmodels)

library(VariantAnnotation)
library(GenomicFeatures)


# Load data
allele_counts_raw <- vroom("~/data/from/popoolation/Sync")

############################### Calculate Allele Frequencies

popsN=7
allele_counts_raw[,c(1:3,4:5,8,9:(9+popsN))] -> allele_counts

names(allele_counts)= c("chr","pos","rc","allele_count","allele_states","major_alleles.maa","minor_alleles.mia","maa_1","maa_2","maa_3","maa_4","maa_5","maa_6","maa_7")

allele_counts  %>%  separate(major_alleles.maa, into = c("MEHILC_ma", "RIDLN_ma", "ICE_ma", "UKW_ma", "NOR_ma", "WCAN_ma", "CAR_ma"), sep = c(1,2,3,4,5,6), remove = F) %>% .[which(.$CAR_ma != "N"),] %>% separate(minor_alleles.mia, into = c("MEHILC_mi", "RIDLN_mi", "ICE_mi", "UKW_mi", "NOR_mi", "WCAN_mi", "CAR_mi"), sep = c(1,2,3,4,5,6), remove = F) %>% .[which(.$CAR_mi == "N"),] %>% separate(maa_1, into = c("MEHILC_count","MEHILC_cov"), sep="/") %>% separate(maa_2, into = c("RIDLN_count","RIDLN_cov"), sep="/") %>% separate(maa_3, into = c("ICE_count","ICE_cov"), sep="/") %>% separate(maa_5, into = c("NOR_count","NOR_cov"), sep="/")  %>% separate(maa_4, into = c("UKW_count","UKW_cov"), sep="/") %>% separate(maa_6, into = c("WCAN_count","WCAN_cov"), sep="/")  %>% separate(maa_7, into = c("CAR_count","CAR_cov"), sep="/") -> allele_counts_sep

Coding_data <- read.delim("~/Functional/Annotation/data", header=T) # this file is generated by running VariantAnnotator using the reference genome and the GFF file.

left_join(allele_counts_sep, Coding_data[,c("chr", "pos","GENEID", "CONSEQUENCE", "REFCODON", "VARCODON", "REFAA", "VARAA")]) -> data_in #Those are the default headers from VariantAnnotator

data_in$CONSEQUENCE = as.character(data_in$CONSEQUENCE)

data_in$CONSEQUENCE[which(data_in$CONSEQUENCE == "synonymous")] = "1.Synonymous" 
data_in$CONSEQUENCE[which(data_in$CONSEQUENCE == "nonsynonymous")] = "2.Non-Synonymous" 
data_in$CONSEQUENCE[which(data_in$CONSEQUENCE == "nonsense")] = "3.Non-Sense"
data_in$CONSEQUENCE[which(data_in$CONSEQUENCE == "not translated")] = "4.Non-Translated" 
data_in$CONSEQUENCE[is.na(data_in$CONSEQUENCE)] = "5.NonCoding" 


counts = c("MEHILC_count","RIDLN_count","ICE_count","UKW_count","NOR_count","WCAN_count")
covs = c("MEHILC_cov","RIDLN_cov","ICE_cov","UKW_cov","NOR_cov","WCAN_cov")


data_in[counts] = lapply(data_in[counts], as.numeric)
data_in[covs] = lapply(data_in[covs], as.numeric)
lapply(data_in[counts], class)
lapply(data_in[covs], class)

data_in %>% mutate(MEHILC_p_car = ifelse(data_in$MEHILC_ma == data_in$CAR_ma, print(data_in$MEHILC_count/data_in$MEHILC_cov), print(1- (data_in$MEHILC_count/data_in$MEHILC_cov)))) %>% mutate(RIDLN_p_car = ifelse(data_in$RIDLN_ma == data_in$CAR_ma, print(data_in$RIDLN_count/data_in$RIDLN_cov), print(1- (data_in$RIDLN_count/data_in$RIDLN_cov)))) %>% mutate(ICE_p_car = ifelse(data_in$ICE_ma == data_in$CAR_ma, print(data_in$ICE_count/data_in$ICE_cov), print(1- (data_in$ICE_count/data_in$ICE_cov)))) %>% mutate(UKW_p_car = ifelse(data_in$UKW_ma == data_in$CAR_ma, print(data_in$UKW_count/data_in$UKW_cov), print(1- (data_in$UKW_count/data_in$UKW_cov)))) %>% mutate(NOR_p_car = ifelse(data_in$NOR_ma == data_in$CAR_ma, print(data_in$NOR_count/data_in$NOR_cov), print(1- (data_in$NOR_count/data_in$NOR_cov)))) %>% mutate(WCAN_p_car = ifelse(data_in$WCAN_ma == data_in$CAR_ma, print(data_in$WCAN_count/data_in$WCAN_cov), print(1- (data_in$WCAN_count/data_in$WCAN_cov)))) %>% mutate(MEHILC_p_WCA = ifelse(data_in$MEHILC_ma == data_in$WCAN_ma, print(data_in$MEHILC_count/data_in$MEHILC_cov), print(1- (data_in$MEHILC_count/data_in$MEHILC_cov)))) %>% mutate(RIDLN_p_WCA = ifelse(data_in$RIDLN_ma == data_in$WCAN_ma, print(data_in$RIDLN_count/data_in$RIDLN_cov), print(1- (data_in$RIDLN_count/data_in$RIDLN_cov)))) %>% mutate(ICE_p_WCA = ifelse(data_in$ICE_ma == data_in$WCAN_ma, print(data_in$ICE_count/data_in$ICE_cov), print(1- (data_in$ICE_count/data_in$ICE_cov)))) %>% mutate(UKW_p_WCA = ifelse(data_in$UKW_ma == data_in$WCAN_ma, print(data_in$UKW_count/data_in$UKW_cov), print(1- (data_in$UKW_count/data_in$UKW_cov)))) %>% mutate(NOR_p_WCA = ifelse(data_in$NOR_ma == data_in$WCAN_ma, print(data_in$NOR_count/data_in$NOR_cov), print(1- (data_in$NOR_count/data_in$NOR_cov)))) %>% mutate(WCAN_p_WCA = ifelse(data_in$WCAN_ma == data_in$WCAN_ma, print(data_in$WCAN_count/data_in$WCAN_cov), print(1- (data_in$WCAN_count/data_in$WCAN_cov)))) %>% mutate(MEHILC_p_rc = ifelse(data_in$MEHILC_ma == data_in$rc, print(data_in$MEHILC_count/data_in$MEHILC_cov), print(1- (data_in$MEHILC_count/data_in$MEHILC_cov)))) %>% mutate(RIDLN_p_rc = ifelse(data_in$RIDLN_ma == data_in$rc, print(data_in$RIDLN_count/data_in$RIDLN_cov), print(1- (data_in$RIDLN_count/data_in$RIDLN_cov)))) %>% mutate(ICE_p_rc = ifelse(data_in$ICE_ma == data_in$rc, print(data_in$ICE_count/data_in$ICE_cov), print(1- (data_in$ICE_count/data_in$ICE_cov)))) %>% mutate(UKW_p_rc = ifelse(data_in$UKW_ma == data_in$rc, print(data_in$UKW_count/data_in$UKW_cov), print(1- (data_in$UKW_count/data_in$UKW_cov)))) %>% mutate(NOR_p_rc = ifelse(data_in$NOR_ma == data_in$rc, print(data_in$NOR_count/data_in$NOR_cov), print(1- (data_in$NOR_count/data_in$NOR_cov)))) %>% mutate(WCAN_p_rc = ifelse(data_in$WCAN_ma == data_in$rc, print(data_in$WCAN_count/data_in$WCAN_cov), print(1- (data_in$WCAN_count/data_in$WCAN_cov)))) -> data_out

############################### Add Ancestral Information

allele_freqs_raw  %>%  separate(`major_alleles(maa)`, into = c("MEHILC_ma", "RIDLN_ma", "ICE_ma", "UKW_ma", "NOR_ma", "WCAN_ma", "CAR_ma"), sep = c(1,2,3,4,5,6), remove = F) %>% .[which(.$CAR_ma != "N"),] %>% separate(`minor_alleles(mia)`, into = c("MEHILC_mi", "RIDLN_mi", "ICE_mi", "UKW_mi", "NOR_mi", "WCAN_mi", "CAR_mi"), sep = c(1,2,3,4,5,6), remove = F) %>% .[which(.$CAR_mi != "N" & .$WCAN_mi != "N"),] -> transpecies_poly

names(transpecies_poly)[1] = "chr"

#cariosus_gt_dat <- read.vcfR("~/data/S_balanoides_genomics_resources/bam_files/single_individuals_DNA/Cariosus/CAR_ind.sbal3.1.snp.vcf.gz")
#cbind(as.data.frame(cariosus_gt_dat@fix), as.data.frame(cariosus_gt_dat@gt)) -> car_gt_df
#vroom_write(car_gt_df, "~/.../cariosus_snp_data.txt")

car_gt_df <- vroom("~/.../cariosus_snp_data.txt") #generated from above


car_gt_df %<>% separate(flt.rmdp.srt.CAR_ind.sbal3.1.bam, into = c("gt","etc"), sep = ":") 

car_gt_df %>% .[which(.$gt == "0/1" & .$QUAL > 100),] -> car_gt_df_filterd
names(car_gt_df_filterd)[1:2] = c("chr","pos")

left_join(transpecies_poly, car_gt_df_filterd[,c("chr","pos","QUAL","gt")] ) -> transpecies_poly

transpecies_poly %>% .[complete.cases(.),] -> transpecies_poly_cc

transpecies_poly_cc %<>% separate(`minor_alleles(mia)`, into = c("N_Atl","Pac","Scar"), sep = c(5,6), remove = F) %>%  mutate( MissingDat_NAtl = str_count(N_Atl,"N")) 
  
transpecies_poly_filter = transpecies_poly_cc[which(transpecies_poly_cc$MissingDat_NAtl %in% c(0,5) ),]

transpecies_poly_filter %>% dim

transpecies_poly_filter %>% .[which(.$N_Atl == "NNNNN"),] %>% dim
